architecture/src/iosMain/kotlin/navigation/Navigator.kt

```diff
@@ -0,0 +1,239 @@
+ * Whenever [navigate] is called, this class maps it to a [NavigationSpec] and performs navigation according to that
+ * @param parent The [UIViewController] managing the navigation
+ * @param navigationMapper A function mapping the [NavigationAction] to [NavigationSpec]
+ */
+actual class Navigator<A : NavigationAction<*>>(parentVC: UIViewController, private val navigationMapper: (A) -> NavigationSpec) {
+
+    private val parent = WeakReference(parentVC)
+
+    actual fun navigate(action: A) {
+        navigate(navigationMapper.invoke(action), action.bundle)
+    }
+
+    private fun navigate(spec: NavigationSpec, bundle: NavigationBundle<*>?) {
+        when (spec) {
+            is NavigationSpec.Push -> pushViewController(spec)
+            is NavigationSpec.Pop -> popViewController(spec)
+            is NavigationSpec.Present -> presentViewController(spec)
+            is NavigationSpec.Dismiss -> dismissViewController(spec)
+            is NavigationSpec.Show -> showViewController(spec)
+            is NavigationSpec.Segue -> segueToViewController(spec, bundle)
+            is NavigationSpec.Nested -> embedNestedViewController(spec)
+            is NavigationSpec.ImagePicker -> presentImagePicker(spec)
+            is NavigationSpec.Email -> presentMailComposer(spec)
+            is NavigationSpec.DocumentSelector -> presentDocumentBrowser(spec)
+            is NavigationSpec.Message -> presentMessageComposer(spec)
+            is NavigationSpec.Phone -> openDialer(spec)
+            is NavigationSpec.Settings -> openSettings()
+            is NavigationSpec.Browser -> openBrowser(spec)
+        }
+        // Since navigation often references UIViewControllers they should be freed up to keep ARC working
+        GC.collect()
+    }
+
+    private fun pushViewController(pushSpec: NavigationSpec.Push) {
+        parent.get()?.navigationController?.pushViewController(pushSpec.push(), pushSpec.animated)
+    }
+
+    private fun popViewController(popSpec: NavigationSpec.Pop) {
+        popSpec.to?.let {
+            parent.get()?.navigationController?.popToViewController(it, popSpec.animated)
+        } ?: parent.get()?.navigationController?.popViewControllerAnimated(popSpec.animated)
+    }
+
+    private fun presentViewController(presentSpec: NavigationSpec.Present) {
+        val toPresent = presentSpec.present()
+        toPresent.modalPresentationStyle = presentSpec.presentationStyle
+        toPresent.modalTransitionStyle = presentSpec.transitionStyle
+        parent.get()?.presentViewController(toPresent, presentSpec.animated) { presentSpec.complete?.invoke() }
+    }
+
+    private fun dismissViewController(dismissSpec: NavigationSpec.Dismiss) {
+        parent.get()?.dismissViewControllerAnimated(dismissSpec.animated) { dismissSpec.complete?.invoke() }
+    }
+
+    private fun showViewController(showSpec: NavigationSpec.Show) {
+        val toShow = showSpec.show()
+        val parent = parent.get() ?: return
```

Comment:
I'd put some `assert` here to known what's going on during development

Reply:
Added assert here and on android Navigator